README.md

[![Travis build
status](https://travis-ci.org/dorianps/LINDA.svg?branch=master)](https://travis-ci.org/dorianps/LINDA)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/ed7c8fb5034e40bfbb12ddbf827078ed)](https://www.codacy.com/manual/dorianps/LINDA?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=dorianps/LINDA&amp;utm_campaign=Badge_Grade)
<!--
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/muschellij2/LINDA?branch=master&svg=true)](https://ci.appveyor.com/project/muschellij2/LINDA)
[![Coverage status](https://codecov.io/gh/muschellij2/LINDA/branch/master/graph/badge.svg)](https://codecov.io/gh/muschellij2/LINDA)

-->
<!-- README.md is generated from README.Rmd. Please edit that file -->

# LINDA Package:

`LINDA` is an R package for automatic segmentation of **chronic** stroke
lesions.  
The method is described in [Hum Brain Mapp. 2016
Apr;37(4):1405-21](http://onlinelibrary.wiley.com/doi/10.1002/hbm.23110/abstract).

-----

## Requirements:

  - Linux or Mac (since Oct 2016 ANTsR [can run in Windows
    10](https://github.com/stnava/ANTsR/wiki/Installing-ANTsR-in-Windows-10-\(along-with-FSL,-Rstudio,-Freesurfer,-etc\).))  
  - [R v3.0 or above](http://www.r-project.org/)  
  - The [ANTsR](http://stnava.github.io/ANTsR/) package installed in R
  - A T1-weighted MRI of a patient with (left hemispheric) stroke

-----

## Installation:

##### Method 1 (easy, incomplete)

This method does not require any prep on your side, just paste the lines
below and all required packages will (hopefully) be installed (including
ANTsR). However, MNI transformations are not included because Github
does not accept big files. You can still force on the fly registrations
to MNI (ch2) space by setting `saveMNI=TRUE`. If you frequently need
outputs in MNI space, use Method 2 below.

``` r
install.packages('devtools')
devtools::install_github('dorianps/LINDA', upgrade_dependencies=FALSE)
```

##### Method 2 (complex, complete)

This method includes the MNI transformations, but works only if you have
previously installed ANTsR. To do that, try:

``` r
devtools::install_github('ANTsX/ANTsR')
```

Then download the [latest LINDA release
v0.5.1](https://github.com/dorianps/LINDA/releases/download/0.5.1/LINDA_v0.5.1.zip)
and install from command line:

``` bash

wget https://github.com/dorianps/LINDA/releases/download/0.5.1/LINDA_v0.5.1.zip
unzip LINDA_v0.5.1.zip # this will unzip to LINDA folder
R CMD INSTALL LINDA # install the LINDA folder in R
```

## Run: 

``` r
library(LINDA)
filename = '/path/to/patient/T1.nii.gz'
outputs = linda_predict(filename)
```

If you don’t specify a filename, a file dialog will allow you to choose
the T1 nifti file of the patient.

``` r
outputs = linda_predict()
```

LINDA will run and you will see the timestamp of the various steps.
Results will be saved in a folder named “linda” in the same path where
the T1 is located. The location of these files will be returned in R
(i.e., in the `outputs` variable).

If the “linda” folder contains segmentations from an earlier run,
processing will stop immediately. Use `cache=FALSE` to force overwriting
the old files. If processing is interrupted and restarted, LINDA will
use the existing files in the “linda” folder to start where it was
interrupted.

**Available prediction models:**  
*Currently a model trained on 60 patients from Penn is used. The
internal prediction engine works with 2mm voxel resolution. This does
not mean your images need to be 2mm, any resolution should work.*

-----

### Example:

\*(a somewhat slow example run on a single CPU core)

``` r
21:18 Starting LINDA v0.5.0
21:18 Creating folder: /data/jux/dpustina/LINDA_package/Sample_ABC/linda
21:18 Loading file: ABC_MPRAGE.nii ...
21:18 Loading template...
21:18 Skull stripping... (long process)
21:46 Saving skull stripped files...
21:46 Computing asymmetry mask...
21:50 Saving asymmetry mask...
21:50 1st round of prediction...
21:50     Running registration: SyN
21:57     Feature calculation 
22:02     Lesion segmentation
22:04 Backprojecting prediction...
22:04 Saving prediction...
22:04 2nd round of prediction...
22:04     Running registration: SyN
22:11     Feature calculation 
22:16     Lesion segmentation
22:17 Backprojecting prediction...
22:17 3rd round of prediction...
22:17     Running registration: SyNCC
00:44     Feature calculation 
00:49     Lesion segmentation
00:50 Backprojecting prediction...
00:50 Saving 3rd final prediction in native space...
00:51 Saving probabilistic prediction in template space...
00:51 Saving probabilistic prediction in native space...
00:51 Skipping data transformation in MNI (ch2) ...
00:51 Done! 3.5 hours 
```

Wonder what to expect? Check individual results from our [60 patients
Penn
dataset](https://drive.google.com/file/d/0BxHeqEv37qqDT085MHAyMzFJcVk)
and [45 patients Georgetown
dataset](https://drive.google.com/open?id=0BxHeqEv37qqDY1psaC14QXZSOXc).  
Our users have reported quite good lesion segmentations for typical
large stroke lesions, but less accurate segmentations with tiny lesions.

-----

##### OUTPUT files:

**BrainMask.nii.gz** - Brain mask used to skull strip (native space)  
**N4corrected.nii.gz** - Bias corrected, full image (native space)  
**N4corrected\_Brain.nii.gz** - Bias corrected, brain only (native
space)  
**N4corrected\_Brain\_LRflipped.nii.gz** - Flipped image used to compute
asymmetry mask (native space)  
**Mask.lesion(1)(2)(3).nii.gz** - masks used for registrations (native
space)  
**Prediction(1)(2)(3).nii.gz** - lesion predictions after each iteration
(template space, but 2mm)  
**Prediction3\_template.nii.gz** - final prediction (template space)  
**Prediction3\_native.nii.gz** - final prediction (native space)  
**Prediction3\_probability\_template.nii.gz** - final graded probability
(template space)  
**Prediction3\_probability\_native.nii.gz** - final graded probability
(native space)  
**Reg3\_registered\_to\_template.nii.gz** - Subject’s MRI, skull
stripped, bias corrected, registered to template (template space)  
**Reg3\_sub\_to\_template\_(affine)(warp)** - transformation matrices to
register subject to template  
**Reg3\_template\_to\_sub\_(affine)(warp)** - transformation matrices to
backproject template to subject  
**Subject\_in\_MNI.nii.gz** - Subject’s MRI, skull stripped, bias
corrected, transformed in MNI space  
**Lesion\_in\_MNI.nii.gz** - Lesion mask in MNI spacethe
**Console\_Output.txt** - log file replicating the console output
**Session\_Info.txt** - R environment and package versions, useful if
you want to reproduce the results in the future.

-----

**Important Note**  
MNI is a space, not a template. There are many templates in MNI space,
most of which do not have the same gyri or sulci at the same
coordinates. LINDA uses the CH2 template which is included in LINDA.
Please don’t use other MNI templates to overlay results. They may look
good from a first look, but they will be wrong. Use the CH2 template
that comes with LINDA. In alternative, you can register the CH2 template
to another template (i.e., ICBM 2009a) and transform back and forth the
results as necessary.

-----

## Frequently Asked Questions

**- What file formats are accepted**?  
Nifti images (.nii and .nii.gz) are accepted. Earlier LINDA versions did
not accept other formats, but now any format can be read. Note that the
Analyze format has unclear left-right orientation.  
**- Can I use it with right hemispheric lesions?**  
Yes, but you need to flip the L-R orientation before. After that, the
prediction will work just as well.  
**- Can I use it with bilateral lesions?**  
It will likely be less accurate. One of the features LINDA uses is the
left-right signal asymmetry.  
**- Can I use it to segment acute and subacute stroke lesions?**  
No, the current model is trained only on chronic stroke patients. It
might be possible to segment acute stroke with models trained on acute
data.  
**- Can I use other images: FLAIR, T2, DWI?**  
No, the existing model accepts only T1w. In principle, additional models
can be built from other modalities (T2, FLAIR, DWI).  
**- Can I train a model with my own data?**  
This is perfectly doable, but the training script is not available
online (needs some work to adapt it for widepsread use). If you want the
example script used for the current model, I can send it easily, just
contact me.  
**- Can I use LINDA to obtain registrations in MNI?**  
The transfer in MNI is obtained by concatenating transformations
“Subject” -\> “Penn Template” -\> “ch2 MNI template”. Thus there are
two sources of potential error. The second source of error is fixed for
all subjects because our template has always the same registration on
MNI. However, a fixed error is always an error. If you really want
precise registration on MNI, I advise to register directly the subject
to an MNI template (possibly using a group MNI template rather than the
ch2).  
**- Will you maintain LINDA and publish new models in the future?**  
There is no plan, time, or funding to do this at this time. If other
researchers want to contribute, this can be done easily because LINDA is
open source.

## Support:

The best way to keep track of bugs or failures is to open a [New
Issue](https://github.com/dorianps/LINDA/issues) on the Github system.
If the algorithm proceeds without errors but the automatic segmentation
is erroneous, please send (i) your T1 image and (ii) the segmentation
produced by LINDA in native space. Try also overlaying
`Mask.lesion*.nii.gz` files on the T1 to check whether the brain mask is
wrong somewhere.


### Authors

Dorian Pustina  
John Muschelli

## Other software for lesion studies

Check out the LESYMAP package for lesion to symptom mapping:
<https://github.com/dorianps/LESYMAP>.


NEWS.md
# LINDA 0.5.1 (2019-09-06)
*   Fixed bug that produced error for existing linda runs

# LINDA 0.5.0
*   Full rewrite to make an R package
*   Use existing outputs as cache to continue interrupted runs
*   Enabled 'cache' argument to overwrite if necessary
*   Github repository sufficient for R package install
*   Console output and other R info saved in output folder
*   Enabled travis build status
*   Enabled automatic check of new versions onAttach
*   Tested on ANTsR versions of late 2017 and late 2018

# LINDA 0.2.7 
*   using splitMask for compatibility with new ANTsR
reconfiguration of June 2017.
*   Added a `NEWS.md` file to track changes to the package.

# LINDA 0.2.6 
*   fixed bug in mrvnrfs_chunks.predict after dimfix from 
@jeffduda in ANTsR (03/02/2017) removed dynamic set of reflaxis, 
all = 0 now

# LINDA 0.2.5 
*   fixed scriptdir to allow command line call

# LINDA 0.2.4 
*   switching mask.lesion1 from graded to binary

# LINDA 0.2.3 
*   fix bug in 'relfaxis'

# LINDA 0.2.2
*   fix axis for left-right refelction

# LINDA 0.2.1 
*   fixed TruncateIntensity issue in old ANTsR
*   added MNI output

# LINDA 0.2.0 
*   added native space output, added probability map output

# LINDA 0.1
*   first published LINDA


linda_predict.R
ver = 'v0.2.7'
cat(paste(format(Sys.time(), "%H:%M") , 'Starting LINDA', ver, "...\n"))
#' Version History
#' 0.1   - first published LINDA
#' 0.2.0 - added native space output, added probability map output
#' 0.2.1 - fixed TruncateIntensity issue in old ANTsR
#'       - added MNI output
#' 0.2.2 - fix axis for left-right refelction
#' 0.2.3 - fix bug in 'relfaxis'
#' 0.2.4 - switching mask.lesion1 from graded to binary
#' 0.2.5 - fixed scriptdir to allow command line call
#' 0.2.6 - fixed bug in mrvnrfs_chunks.predict after dimfix 
#'         from @jeffduda in ANTsR (03/02/2017)
#'         removed dynamic set of reflaxis, all = 0 now
#' 0.2.7 - using splitMask for compatibility with new ANTsR
#'         reconfiguration of June 2017.

# check for necessary packages and load them
if (! is.element("ANTsR", installed.packages()[,1])) {
  stop("Required ANTsR package cannot be found.
        Automated installation not possible.
       See http://stnava.github.io/ANTsR for installation instructions")
} else {
  library(ANTsR)
}
if (! is.element("randomForest", installed.packages()[,1])) {
  print("Installing missing `randomForest` package")
  install.packages("randomForest")
  library(randomForest)
} else { 
  library(randomForest)
}


# parse arguments for command line use
args=commandArgs(TRUE)
for(i in 1:length(args)){
  eval(parse(text=args[i]))
}


# select the file
if (! exists('t1')) {
  t1 = file.choose()
} else {
  if (! file.exists(t1)) {
    stop(paste( 'File inexistent:', t1 ))
  }
}

                                                  
# check this is a nifti extension
if (length(grep(".nii$", t1)) == 0 & length(grep(".nii.gz$", t1)) == 0) {
  stop(
    paste( 'Unsupported file type ->', basename(t1) )
  )
}


# load the file
cat(paste(format(Sys.time(), "%H:%M") , 'Loading file:', basename(t1), "\n"))
subimg = antsImageRead(t1)
submask = subimg*0 + 1


# create linda folder
lindadir = file.path(dirname(t1), 'linda')
cat(paste(format(Sys.time(), "%H:%M") , 'Creating folder:', lindadir, "\n"))
dir.create(lindadir, showWarnings = F, recursive = T)


# fix TruncateIntensity incompatibility with old ANTsR binaries
if ( length( grep('TruncateIntensity' ,iMath(20,'GetOperations'))) != 0 ) {
  truncate = 'TruncateIntensity'
} else {
  truncate = 'TruncateImageIntensity'
}


# load other functions
if (!exists('scriptdir'))  scriptdir = dirname(sys.frame(1)$ofile)
source(file.path(scriptdir, 'getLesionFeatures.R'), echo=F)
source(file.path(scriptdir, 'mrvnrfs_chunks.R'), echo=F)


# load template files
cat(paste(format(Sys.time(), "%H:%M") , "Loading template... \n"))
temp = antsImageRead(file.path(scriptdir,'pennTemplate','template.nii.gz'))
tempbrain = antsImageRead(file.path(scriptdir,'pennTemplate','templateBrain.nii.gz'))
tempmask = antsImageRead(file.path(scriptdir,'pennTemplate','templateBrainMask.nii.gz'))



# two rounds of N4-BrainExtract to skull strip
cat(paste(format(Sys.time(), "%H:%M") , "Skull stripping... (long process) \n"))
for (i in 1:2) {
  n4 = abpN4(img = subimg, mask=submask)
  bextract = abpBrainExtraction(img = subimg, 
                                tem = temp,
                                temmask = tempmask)
  submask = bextract$bmask*1
}
simg = n4 * submask
cat(paste(format(Sys.time(), "%H:%M") , "Saving skull stripped files \n"))
antsImageWrite(n4, file.path(lindadir,'N4corrected.nii.gz'))
antsImageWrite(submask, file.path(lindadir,'BrainMask.nii.gz'))
antsImageWrite(simg, file.path(lindadir,'N4corrected_Brain.nii.gz'))

cat(paste(format(Sys.time(), "%H:%M") , "Loading LINDA model \n"))
load(file.path(scriptdir, 'PublishablePennModel_2mm_mr321_rad1.Rdata'))


# compute asymmetry mask
cat(paste(format(Sys.time(), "%H:%M") , "Computing asymmetry mask... \n"))
reflaxis = 0 # which.max(abs(antsGetDirection(simg)[,1]))-1

asymmetry = reflectImage(simg,axis=reflaxis, tx='Affine'); Sys.sleep(2)
antsImageWrite(asymmetry$warpedmovout, file.path(lindadir,'N4corrected_Brain_LRflipped.nii.gz'))
reflect = smoothImage(asymmetry$warpedmovout, 2) - smoothImage(simg, 2)
reflect[reflect<0]=0
reflect = iMath(reflect,'Normalize')
mask.lesion1 = submask - reflect
mask.lesion1 = thresholdImage(mask.lesion1,0.6,Inf) * submask
cat(paste(format(Sys.time(), "%H:%M") , "Saving asymmetry mask... \n"))
antsImageWrite(mask.lesion1, file.path(lindadir,'Mask.lesion1_asym.nii.gz'))


# half brain mask
resamplevox = c(2,2,2)
brainmask=iMath(tempmask, 'MD', 1)
emptyimg = brainmask*1; emptyimg=as.array(emptyimg); emptyimg[ 1:91 , ,]=0
brainmaskleftbrain = resampleImage(brainmask*emptyimg, resamplevox, 0, 1) 
rm(emptyimg)


################################ 1st registration 
cat(paste(format(Sys.time(), "%H:%M") , "Running 1st registration... \n"))
reg1=antsRegistration(fixed=simg,moving=tempbrain,typeofTransform = 'SyN',mask=mask.lesion1)
reg1$warpedfixout = reg1$warpedfixout %>% 
  iMath(truncate,0.01,0.99) %>% 
  iMath('Normalize')
tempmask=antsApplyTransforms(moving=submask, fixed=tempbrain,transformlist = reg1$invtransforms, interpolator = 'NearestNeighbor')

# prepare features
cat(paste(format(Sys.time(), "%H:%M") , "Feature calculation... \n"))
features = getLesionFeatures(reg1$warpedfixout, tempbrain, scriptdir,tempmask,truncate,reflaxis)
for (i in 1:length(features)) features[[i]] = resampleImage(features[[i]], resamplevox, 
                                                            useVoxels = 0, interpType = 0) * brainmaskleftbrain

# 1st prediction
cat(paste(format(Sys.time(), "%H:%M") , "Running 1st prediction... \n"))
predlabel.sub = brainmaskleftbrain*1;
predlabel.sub[features[[4]] == 0] = 0
mmseg1<-suppressMessages(
  mrvnrfs.predict_chunks( rfm$rflist, list(features),
                               predlabel.sub, rad=rad,
                               multiResSchedule=mr, voxchunk=5000 )
)
cat(paste(format(Sys.time(), "%H:%M") , "Saving 1st prediction... \n"))
antsImageWrite(mmseg1$seg[[1]], file.path(lindadir,'Prediction1.nii.gz'))

# backproject
cat(paste(format(Sys.time(), "%H:%M") , "Backprojecting 1st prediction... \n"))
seg = resampleImage(mmseg1$seg[[1]], dim(tempbrain), useVoxels = 1, interpType = 1)
seg[seg!=4]=0
seg[seg==4]=1
segnative = antsApplyTransforms(fixed = simg, moving = seg, 
                                transformlist = reg1$fwdtransforms, interpolator = 'NearestNeighbor')
mask.lesion2=submask*1
mask.lesion2[segnative==1]=0
antsImageWrite(mask.lesion2, file.path(lindadir,'Mask.lesion2.nii.gz'))



########################### 2nd registration 
cat(paste(format(Sys.time(), "%H:%M") , "Running 2nd registration... \n"))
reg2=antsRegistration(fixed=simg,moving=tempbrain,typeofTransform = 'SyN',mask=mask.lesion2)
reg2$warpedfixout = reg2$warpedfixout %>% 
  iMath(truncate,0.01,0.99) %>% 
  iMath('Normalize')
tempmask=antsApplyTransforms(moving=submask, fixed=tempbrain,transformlist = reg2$invtransforms, interpolator = 'NearestNeighbor')

# prepare features
cat(paste(format(Sys.time(), "%H:%M") , "Feature calculation... \n"))
features = getLesionFeatures(reg2$warpedfixout, tempbrain, scriptdir, tempmask,truncate,reflaxis)
for (i in 1:length(features)) features[[i]] = resampleImage(features[[i]], resamplevox, 
                                                            useVoxels = 0, interpType = 0) * brainmaskleftbrain

# 2nd prediction
cat(paste(format(Sys.time(), "%H:%M") , "Running 2nd prediction... \n"))
predlabel.sub = brainmaskleftbrain*1;
predlabel.sub[features[[4]] == 0] = 0
mmseg2<-suppressMessages(
  mrvnrfs.predict_chunks( rfm$rflist, list(features),
                                predlabel.sub, rad=rad,
                                multiResSchedule=mr, voxchunk=5000 )
)
cat(paste(format(Sys.time(), "%H:%M") , "Saving 2nd prediction... \n"))
antsImageWrite(mmseg2$seg[[1]], file.path(lindadir,'Prediction2.nii.gz'))

# backproject
cat(paste(format(Sys.time(), "%H:%M") , "Backprojecting 2nd prediction... \n"))
seg = resampleImage(mmseg2$seg[[1]], dim(tempbrain), useVoxels = 1, interpType = 1)
seg[seg!=4]=0
seg[seg==4]=1
segnative = antsApplyTransforms(fixed = simg, moving = seg, 
                                transformlist = reg2$fwdtransforms, interpolator = 'NearestNeighbor')
mask.lesion3=submask*1
mask.lesion3[segnative==1]=0
antsImageWrite(mask.lesion3, file.path(lindadir,'Mask.lesion3.nii.gz'))



########################### 3rd registration 
cat(paste(format(Sys.time(), "%H:%M") , "Running 3rd registration... (long process)\n"))
reg3=antsRegistration(fixed=simg,moving=tempbrain,typeofTransform = 'SyNCC',mask=mask.lesion3)
# save reg3 results
cat(paste(format(Sys.time(), "%H:%M") , "Saving 3rd registration results... \n"))
antsImageWrite(reg3$warpedfixout, file.path(lindadir,'Reg3_registered_to_template.nii.gz'))
file.copy(reg3$fwdtransforms[1], file.path(lindadir , 'Reg3_template_to_sub_warp.nii.gz'))
file.copy(reg3$fwdtransforms[2], file.path(lindadir , 'Reg3_template_to_sub_affine.mat'))
file.copy(reg3$invtransforms[1], file.path(lindadir , 'Reg3_sub_to_template_affine.mat'))
file.copy(reg3$invtransforms[2], file.path(lindadir , 'Reg3_sub_to_template_warp.nii.gz'))

reg3$warpedfixout = reg3$warpedfixout %>% 
  iMath(truncate,0.01,0.99) %>% 
  iMath('Normalize')
tempmask=antsApplyTransforms(moving=submask, fixed=tempbrain,transformlist = reg3$invtransforms, interpolator = 'NearestNeighbor')

# prepare features
cat(paste(format(Sys.time(), "%H:%M") , "Feature calculation... \n"))
features = getLesionFeatures(reg3$warpedfixout, tempbrain, scriptdir, tempmask,truncate,reflaxis)
for (i in 1:length(features)) features[[i]] = resampleImage(features[[i]], resamplevox, 
                                                            useVoxels = 0, interpType = 0) * brainmaskleftbrain

# 3rd prediction
cat(paste(format(Sys.time(), "%H:%M") , "Running 3rd prediction... \n"))
predlabel.sub = brainmaskleftbrain*1;
predlabel.sub[features[[4]] == 0] = 0
mmseg3<-suppressMessages(
  mrvnrfs.predict_chunks( rfm$rflist, list(features),
                                predlabel.sub, rad=rad,
                                multiResSchedule=mr, voxchunk=5000 )
)
cat(paste(format(Sys.time(), "%H:%M") , "Saving 3rd final prediction... \n"))
antsImageWrite(mmseg3$seg[[1]], file.path(lindadir,'Prediction3.nii.gz'))


# backproject to save
seg = resampleImage(mmseg3$seg[[1]], dim(tempbrain), useVoxels = 1, interpType = 1)
seg[seg!=4]=0
seg[seg==4]=1
segnative = antsApplyTransforms(fixed = simg, moving = seg, 
                                transformlist = reg3$fwdtransforms, interpolator = 'NearestNeighbor')
cat(paste(format(Sys.time(), "%H:%M") , "Saving 3rd final prediction in template space... \n"))
antsImageWrite(seg, file.path(lindadir,'Prediction3_template.nii.gz'))
cat(paste(format(Sys.time(), "%H:%M") , "Saving 3rd final prediction in native space... \n"))
antsImageWrite(segnative, file.path(lindadir,'Prediction3_native.nii.gz'))

# save graded map
probles = resampleImage(mmseg3$probs[[1]][[4]], dim(tempbrain), useVoxels = 1, interpType = 0)
problesnative = antsApplyTransforms(fixed = simg, moving = probles, 
                                transformlist = reg3$fwdtransforms, interpolator = 'Linear')
cat(paste(format(Sys.time(), "%H:%M") , "Saving probabilistic prediction in template space... \n"))
antsImageWrite(probles, file.path(lindadir,'Prediction3_probability_template.nii.gz'))
cat(paste(format(Sys.time(), "%H:%M") , "Saving probabilistic prediction in native space... \n"))
antsImageWrite(problesnative, file.path(lindadir,'Prediction3_probability_native.nii.gz'))

# save in MNI coordinates
cat(paste(format(Sys.time(), "%H:%M") , "Transferring data in MNI (ch2) space... \n"))
warppenn = file.path(lindadir , 'Reg3_sub_to_template_warp.nii.gz')
affpenn = file.path(lindadir , 'Reg3_sub_to_template_affine.mat')
warpmni = file.path(scriptdir,'pennTemplate','templateToCh2_1Warp.nii.gz')
affmni = file.path(scriptdir,'pennTemplate','templateToCh2_0GenericAffine.mat')
mni = antsImageRead(
    file.path(scriptdir,'pennTemplate','ch2.nii.gz')
  )
matrices = c(warpmni,affmni,affpenn,warppenn)

submni=antsApplyTransforms(moving=simg, fixed=mni,transformlist = matrices, interpolator = 'Linear', whichtoinvert = c(0,0,1,0))
lesmni=antsApplyTransforms(moving=segnative, fixed=mni,transformlist = matrices, interpolator = 'NearestNeighbor', whichtoinvert = c(0,0,1,0))

cat(paste(format(Sys.time(), "%H:%M") , "Saving subject in MNI (ch2) space... \n"))
antsImageWrite(submni, file.path(lindadir,'Subject_in_MNI.nii.gz'))
cat(paste(format(Sys.time(), "%H:%M") , "Saving lesion in MNI (ch2) space... \n"))
antsImageWrite(lesmni, file.path(lindadir,'Lesion_in_MNI.nii.gz'))


cat('DONE')